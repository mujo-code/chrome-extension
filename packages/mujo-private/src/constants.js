// config
export const API_HOST = 'https://getmujo.com'

// storage
export const SITE_TIME_KEY = 'SITE_TIME'
export const PREDICTED_BREAK_TIMES = 'PREDICTED_BREAK_TIMES'

// features
export const SMART_BREAKS_FEATURE = true

// alarms
export const SMART_BREAK = 'SMRT_BRK' // is only part of the key
// ex. SMRT_BRK_2019-08-07T22:17:13.246Z
