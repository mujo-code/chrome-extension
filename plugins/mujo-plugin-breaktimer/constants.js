export const ALARM_KEY = 'MINDFUL_ALARM'
export const PREDICTED_TIMES_FEATURE = 'PREDICTED_BREAK_TIMES_FEATURE'
export const HOUR_HALF = 1000 * 60 * 60 * 1.5
