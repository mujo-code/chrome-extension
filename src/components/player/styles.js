import { css } from '@emotion/core'

export const transition = () => css({ transition: 'all 0.7s' })

export const toSvgProps = ({ width, height, isOpen }, animating) =>
  css(
    {
      width: isOpen ? '100vw' : width,
      height: isOpen ? '100vh' : height,
      viewBox: `0 0 ${width} ${height}`,
    },
    transition({ width, height, isOpen }),
    animating
  )

export const toRectProps = ({ width, height, isOpen }, animating) =>
  css(
    {
      x: 0,
      y: 0,
      width: isOpen ? '100vw' : width,
      height: isOpen ? '100vh' : height,
      rx: isOpen ? 0 : height / 11,
    },
    transition({ width, height, isOpen }),
    animating
  )

export const toCircleProps = (
  { width, height, isOpen, circleRatio },
  animating
) =>
  css(
    {
      cx: isOpen ? '50vw' : width / 2,
      cy: isOpen ? '50vh' : height / 2,
      r: isOpen
        ? window.innerHeight * 0.2 // only use ratio for icon
        : height * circleRatio,
      transform: 'scale(1) translateZ(0)',
      transformOrigin: 'center',
    },
    transition({ width, height, isOpen }),
    animating
  )

export const toTextProps = ({ width, height, isOpen }, animating) =>
  Object.assign(
    {
      x: isOpen ? window.innerWidth / 2 : width / 2,
      // NOTE: goes down half font size
      y: (isOpen ? window.innerHeight / 2 : height / 2) + 8,
      textAnchor: 'middle',
    },
    css(
      { width: '300px', textAlign: 'center' },
      transition({ width, height, isOpen }),
      animating
    )
  )

export const toCountProps = ({ width, height, isOpen }, animating) =>
  Object.assign(
    {
      x: isOpen ? window.innerWidth / 2 : width / 2,
      // NOTE: goes down half font size
      y: (isOpen ? window.innerHeight : height) - 16,
      textAnchor: 'middle',
    },
    css(
      { width: '300px', textAlign: 'center' },
      transition({ width, height, isOpen }),
      animating
    )
  )
