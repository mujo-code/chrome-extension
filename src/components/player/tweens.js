import * as TWEEN from '@tweenjs/tween.js'

const loop = shouldLoop => time => {
  if (shouldLoop()) {
    requestAnimationFrame(loop(shouldLoop))
    TWEEN.update(time)
  }
}

const defaultsOptions = {
  repeat: 2,
  startDelay: 700,
  circleScale: 1,
  circleScaleExpanded: 1.6,
  circleInnerScale: 1,
  circleInnerExpanded: 1.4,
  breathTime: 3000,
}

export const createTween = inputOptions => {
  const options = { ...defaultsOptions, ...inputOptions }
  const {
    updateTween,
    completeTween,
    setTween,
    repeat,
    startDelay,
    circleScale,
    circleScaleExpanded,
    circleInnerScale,
    circleInnerExpanded,
    breathTime,
    shouldLoop,
  } = options
  const renderLoop = loop(shouldLoop)
  const position = { scale: circleScale, scale2: circleInnerScale }
  const breathIn = new TWEEN.Tween(position)
    .to(
      { scale: circleScaleExpanded, scale2: circleInnerExpanded },
      breathTime
    )
    .easing(TWEEN.Easing.Back.InOut)
  const update = updateTween({
    tween: breathIn,
    shouldUpdate: shouldLoop,
  })
  const breathOut = new TWEEN.Tween(position)
    .to({ scale: circleScale, scale2: circleInnerScale }, breathTime)
    .easing(TWEEN.Easing.Back.InOut)
    .onUpdate(update)
    .onComplete(completeTween({ tween: breathIn, ...options }))

  breathIn.chain(breathOut).onUpdate(update)

  // NOTE: start every thing
  const timer = setTimeout(() => {
    breathIn.start()
  }, startDelay)
  requestAnimationFrame(renderLoop)

  // return function to teardown
  return {
    repeat,
    update: ({
      updateTween: nextUpdateTween,
      shouldLoop: nextShouldLoop,
      completeTween: nextCompleteTween,
    }) => {
      const nextUpdate = nextUpdateTween({
        tween: breathIn,
        shouldUpdate: nextShouldLoop,
      })
      breathIn.onUpdate(nextUpdate)
      breathOut.onUpdate(nextUpdate)
      breathOut.onComplete(
        nextCompleteTween({ tween: breathIn, ...options })
      )
    },
    stop: () => {
      setTween(null)
      clearTimeout(timer)
      breathIn.stop()
      TWEEN.removeAll()
    },
  }
}
