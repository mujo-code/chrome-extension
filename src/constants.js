// Config values
export const MINUTE = 1000 * 60
export const FIVE_MINUTES = MINUTE * 5
export const FOURTY_FIVE_MINUTES = MINUTE * 45
export const HOUR = MINUTE * 60
export const THREE_HOURS = HOUR * 3
export const ALARM_DEFAULT_VALUE = true
export const MAX_BREAKTIMERS = 5
export const CURRENT_SUB_SKU = 'early_subscription'
export const APP_READY_KEY = '__mujo_ready__'

// ALARMS
export const ALARM_KEY = 'MINDFUL_ALARM' // also storage key
export const HEARTBEAT = 'HEARTBEAT'

// DB Keys
export const DATABASE_NAME = 'MUJO'
export const DATABASE_STORE = 'STATS'
export const DATABASE_VERSION = 1
export const LAST_ACTIVITY_TABLE = 'LAST_ACTIVITY'

// Storage keys
export const TOP_SITES_KEY = 'TOP_SITES'
export const TOP_SITES_USAGE_KEY = 'TOP_SITES_USAGE'
export const PAGE_VIEWS_KEY = 'PAGE_VIEWS'
export const SHOW_TOP_SITES_KEY = 'SHOW_TOP_SITES'
export const SITE_TIME_KEY = 'SITE_TIME'
export const BREAK_TIMERS_KEY = 'BREAK_TIMERS'
export const LAST_ACTIVE_KEY = 'LAST_ACTIVE_KEY'
export const ACTIVITY_NUMBER_KEY = 'ACTIVITY_NUMBER_KEY'

// Actions
export const NEW_TAB_CONNECTION = 'New Connection'
export const PAGE_VIEWING_TIME = 'Page Viewing Time'
export const DEEP_LINK_NEWTAB = 'Deep Link New Tab'
export const RESET_USAGE = 'Reset Usage Stats'
export const TRACK = 'Track'
export const ADD_BROADCAST_TAB = 'Add Broadcast Tab'
export const REMOVE_BROADCAST_TAB = 'Remove Broadcast Tab'
export const VALUE_CHANGED = 'Value changed'

// Getter, Setters for storage
export const GET_STORAGE = 'GET_STORAGE'
export const SET_STORAGE = 'SET_STORAGE'

// Feature Flags
export const SCREEN_TIME_FEATURE = true
export const BREAK_TIMER_FEATURE = true
export const SUBSCRIBE_FEATURE = false

// Upsell modals
export const MAX_BREAKTIMER_MODAL = 'breakTimerMax'
export const SUB_DETAILS_MODAL = 'subscriptionDetails'
export const SUB_SUCCESS_MODAL = 'subSuccess'

// OPTIONAL permissions requested
export const SCREEN_TIME_PERMISSIONS = {
  permissions: ['activeTab'],
  origins: ['https://*/*', 'http://*/*'],
}
