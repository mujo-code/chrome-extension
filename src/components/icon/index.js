import { Box } from '@mujo/box'
import { removeKeys } from '@mujo/box/dist/lib/remove-keys'
import React from 'react'
import { Menu, X } from 'react-feather'
import { colors } from '../../styles/colors'

// maintain list to avoid blowing up bundle size
// if using another icon please add to icons list
const icons = { menu: Menu, close: X }
const sizes = { s: 16, m: 24, l: 32 }
const NotFound = ({ icon }) => (
  <Box>{console.warn(`Warning icon ${icon} not found`) && null}</Box>
)

export const Icon = props => {
  const { icon, color, size } = props
  const otherProps = removeKeys(props, 'icon', 'color', 'size')
  const Component = icons[icon] || NotFound
  return (
    <Box {...otherProps}>
      <Component
        icon={icon}
        color={colors[color] || colors.saltBox}
        size={sizes[size] || sizes.m}
      />
    </Box>
  )
}
