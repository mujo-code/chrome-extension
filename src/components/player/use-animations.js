import { useEffect, useState, useCallback } from 'react'
import {
  toSvgProps,
  toCircleProps,
  toTextProps,
  toRectProps,
  toCountProps,
} from './styles'
import { createTween } from './tweens'

export { transition } from './styles'

const REPEAT = 5
export const wait = fn => (...args) =>
  setTimeout(() => fn(...args), 0)

const animatingPropsFactory = () => ({
  rect: {},
  circle: {},
  circle2: {},
})

const defaultsOptions = {
  breathTimeIncrease: 300,
  circleScaleExpandedIncrease: 0.2,
  circleInnerExpandedIncrease: 0.1,
}

export const useAnimations = (props, argOptions = {}) => {
  const options = Object.assign({}, defaultsOptions, argOptions)
  const { isOpen, onFinish } = props
  const [animatingProps, setAnimatingProps] = useState(
    animatingPropsFactory()
  )
  const [isBreathIn, setIsBreathIn] = useState(true)
  const [tween, setTween] = useState(null)
  const delayedSetInBreath = wait(setIsBreathIn)
  const delayedSetAnimatingProps = wait(setAnimatingProps)
  const updateTween = useCallback(
    ({ tween: currentTween, shouldUpdate }) => ({
      scale,
      scale2,
    }) => {
      /* eslint-disable-next-line no-underscore-dangle */
      const isPlaying = currentTween._isPlaying
      if (isBreathIn !== isPlaying) {
        delayedSetInBreath(isPlaying)
      }
      if (!shouldUpdate()) {
        return
      }
      setAnimatingProps(
        Object.assign({}, animatingProps, {
          circle: {
            transform: `scale(${scale2})`,
            transition: 'none',
          },
          circle2: {
            transform: `scale(${scale})`,
            transition: 'all 0.1s',
          },
        })
      )
    },
    [animatingProps, delayedSetInBreath, isBreathIn]
  )
  // TODO: update to be able to hot swappable like update
  const completeTween = useCallback(
    ({
      repeat: iteration,
      breathTime,
      circleScaleExpanded,
      circleInnerExpanded,
    }) => () => {
      if (iteration > 1) {
        const {
          breathTimeIncrease,
          circleScaleExpandedIncrease,
          circleInnerExpandedIncrease,
        } = options
        setTween(
          createTween({
            updateTween,
            completeTween,
            startDelay: 0,
            repeat: iteration - 1,
            shouldLoop: () => isOpen,
            breathTime: breathTime + breathTimeIncrease,
            circleScaleExpanded:
              circleScaleExpanded + circleScaleExpandedIncrease,
            circleInnerExpanded:
              circleInnerExpanded + circleInnerExpandedIncrease,
            setTween,
          })
        )
        return
      }
      setAnimatingProps(animatingPropsFactory())
      onFinish()
    },
    [isOpen, onFinish, options, updateTween]
  )
  useEffect(() => {
    if (isOpen && !tween) {
      setTween(
        createTween({
          updateTween,
          completeTween,
          repeat: REPEAT,
          shouldLoop: () => isOpen,
          setTween,
        })
      )
    } else if (tween && !isOpen) {
      tween.stop()
      delayedSetAnimatingProps(
        Object.assign({}, animatingProps, {
          circle: {},
          circle2: {},
        })
      )
    } else if (tween) {
      tween.update({
        updateTween,
        shouldLoop: () => isOpen,
        completeTween,
      })
    }
  }, [
    isOpen,
    animatingProps,
    isBreathIn,
    tween,
    updateTween,
    completeTween,
    delayedSetAnimatingProps,
  ])
  const animationProps = {
    svg: toSvgProps(props),
    rect: toRectProps(props, animatingProps.rect),
    circle: toCircleProps(props, animatingProps.circle),
    circle2: toCircleProps(props, animatingProps.circle2),
    text: toTextProps(props, {}),
    count: toCountProps(props, {}),
  }
  return [
    {
      iteration: tween ? tween.repeat : 0,
      animationProps,
      isBreathIn,
    },
    {},
  ]
}
