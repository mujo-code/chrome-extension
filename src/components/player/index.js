import { Box } from '@jcblw/box'
import { removeKeys } from '@jcblw/box/dist/lib/remove-keys'
import { css } from 'glamor'
import React from 'react'
import { useTheme } from '../../hooks/use-theme'
import { HeaderL, HeaderS } from '../fonts'
import { useAnimations, transition } from './use-animations'

const fadeInGroup = css({
  transition: 'opacity 0.7s ease-in 0.7s',
  opacity: 0,
  ':not(:empty)': { opacity: 1 },
})

const textTranistions = css({
  transformOrigin: 'center',
  transition: 'all 0.1s ease-out 0s',
  opacity: 0,
  transform: 'scale(0.9)',
})

const fadeInText = css({
  transition: 'all 0.3s ease-in 0.5s',
  opacity: 1,
  transform: 'scale(1)',
})

export const Player = props => {
  const { isOpen, label } = props
  const otherProps = removeKeys(
    props,
    'width',
    'height',
    'isOpen',
    'circleRatio',
    'onFinish',
    'label'
  )
  const [{ animationProps, isBreathIn, iteration }] = useAnimations(
    props,
    isOpen
  )
  const { foreground, background, highlight } = useTheme()
  return (
    <Box
      paddingTop={isOpen ? 'none' : 'm'}
      {...transition(props)}
      {...otherProps}
    >
      <Box
        Component="svg"
        version="1.1"
        xmlns="http://www.w3.org/2000/svg"
        {...animationProps.svg}
        css={{ pointerEvents: 'none' }}
      >
        <Box
          Component="rect"
          fill={background}
          {...animationProps.rect}
        />
        <Box
          Component="circle"
          fill={highlight}
          {...animationProps.circle2}
        />
        <Box
          Component="circle"
          fill={foreground}
          {...animationProps.circle}
        />
        <Box Component="g" {...fadeInGroup}>
          {isOpen ? (
            <>
              <HeaderL
                fill={background}
                Component="text"
                {...animationProps.text}
                {...css(
                  textTranistions,
                  isBreathIn ? fadeInText : {}
                )}
              >
                Breathe in
              </HeaderL>
              <HeaderL
                fill={background}
                Component="text"
                {...textTranistions}
                {...animationProps.text}
                {...css(
                  textTranistions,
                  !isBreathIn ? fadeInText : {}
                )}
              >
                Breathe out
              </HeaderL>
              <HeaderS
                Component="text"
                fill={foreground}
                {...textTranistions}
                {...animationProps.count}
                {...css(textTranistions, fadeInText)}
              >
                {iteration}
              </HeaderS>
            </>
          ) : null}
          <HeaderL
            fill={background}
            Component="text"
            {...textTranistions}
            {...animationProps.text}
            {...css(textTranistions, !isOpen ? fadeInText : {})}
          >
            {label}
          </HeaderL>
        </Box>
      </Box>
    </Box>
  )
}

Player.defaultProps = { circleRatio: 0.2 }
