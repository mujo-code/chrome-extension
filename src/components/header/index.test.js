import React from 'react'
import { create } from 'react-test-renderer'
import { SETTINGS_MODAL } from '../../constants'
import { Icon } from '../icon'
import { Player } from '../player'
import { Header } from '.'

test('Header component matches snapshot', () => {
  const tree = create(<Header />).toJSON()
  expect(tree).toMatchSnapshot()
})

test('Header component matches open snapshot', () => {
  const tree = create(<Header playerIsOpen={true} />).toJSON()
  expect(tree).toMatchSnapshot()
})

test('Header should open settings modal when the Icon is clicked', () => {
  const fn = jest.fn()
  const wrapper = create(<Header setUpsellModal={fn} />)
  const { root } = wrapper
  const icon = root.findByType(Icon)

  // mock click on modal
  icon.props.onClick({ target: icon })
  expect(fn).toBeCalledWith({ name: SETTINGS_MODAL })
})

test('Header should open settings modal when the Icon is clicked', () => {
  const fn = jest.fn()
  const wrapper = create(<Header setUpsellModal={fn} />)
  const { root } = wrapper
  const icon = root.findByType(Icon)

  // mock click on modal
  icon.props.onClick({ target: icon })
  expect(fn).toBeCalledWith({ name: SETTINGS_MODAL })
})

test('Clicking the Header player should open it', () => {
  const fn = jest.fn()
  const wrapper = create(<Header setPlayerIsOpen={fn} />)
  const { root } = wrapper
  const player = root.findByType(Player)

  // mock click on modal
  player.props.onClick({ target: player })
  expect(fn).toBeCalledWith(true)
})
