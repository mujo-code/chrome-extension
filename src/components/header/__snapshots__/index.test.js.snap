// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header component matches open snapshot 1`] = `
<div
  className="css-1s1haim"
  layer="2"
  position="relative"
>
  <div
    className="css-1cq7tb4"
    onClick={[Function]}
    position="absolute"
  >
    <svg
      fill="none"
      height={32}
      icon="menu"
      stroke="#EAE2EB"
      strokeLinecap="round"
      strokeLinejoin="round"
      strokeWidth="2"
      viewBox="0 0 24 24"
      width={32}
      xmlns="http://www.w3.org/2000/svg"
    >
      <line
        x1="3"
        x2="21"
        y1="12"
        y2="12"
      />
      <line
        x1="3"
        x2="21"
        y1="6"
        y2="6"
      />
      <line
        x1="3"
        x2="21"
        y1="18"
        y2="18"
      />
    </svg>
  </div>
  <div
    className="css-t94bnt"
    data-testid="player"
    onClick={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
  >
    <svg
      className="css-ui4khd"
      version="1.1"
      xmlns="http://www.w3.org/2000/svg"
    >
      <rect
        className="css-1j7or71"
        fill="outerSpace"
      />
      <circle
        className="css-glnwju"
        fill="saltBox"
      />
      <circle
        className="css-glnwju"
        fill="mischka"
      />
      <g
        className="css-1itd6tp"
      >
        <text
          className="css-pdmncd"
          color="color"
          fill="outerSpace"
          textAnchor="middle"
          x={512}
          y={392}
        >
          Breathe in
        </text>
        <text
          className="css-13elofn"
          color="color"
          fill="outerSpace"
          textAnchor="middle"
          x={512}
          y={392}
        >
          Breathe out
        </text>
        <text
          className="css-1tprw8m"
          color="color"
          fill="mischka"
          textAnchor="middle"
          x={512}
          y={752}
        >
          0
        </text>
        <text
          className="css-1vhvvnq"
          color="color"
          fill="outerSpace"
          textAnchor="middle"
          x={512}
          y={392}
        />
      </g>
    </svg>
  </div>
  <div
    className="css-0"
    position="relative"
  >
    <div
      className="css-1cncp8y"
      position="absolute"
    >
      <span
        backgroundColor="mischka"
        className="css-1bxc77d"
        color="outerSpace"
        position="relative"
      >
        <div
          backgroundColor="mischka"
          className="css-34aj3n"
          position="absolute"
        />
        <div
          className="css-4obz46"
          position="relative"
        >
          Take a break!
        </div>
      </span>
    </div>
  </div>
</div>
`;

exports[`Header component matches snapshot 1`] = `
<div
  className="css-1s1haim"
  layer="2"
  position="relative"
>
  <div
    className="css-z2kty0"
    onClick={[Function]}
    position="absolute"
  >
    <svg
      fill="none"
      height={32}
      icon="menu"
      stroke="#EAE2EB"
      strokeLinecap="round"
      strokeLinejoin="round"
      strokeWidth="2"
      viewBox="0 0 24 24"
      width={32}
      xmlns="http://www.w3.org/2000/svg"
    >
      <line
        x1="3"
        x2="21"
        y1="12"
        y2="12"
      />
      <line
        x1="3"
        x2="21"
        y1="6"
        y2="6"
      />
      <line
        x1="3"
        x2="21"
        y1="18"
        y2="18"
      />
    </svg>
  </div>
  <div
    className="css-pjuhn5"
    data-testid="player"
    onClick={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
  >
    <svg
      className="css-63egd1"
      version="1.1"
      xmlns="http://www.w3.org/2000/svg"
    >
      <rect
        className="css-15od96u"
        fill="outerSpace"
      />
      <circle
        className="css-1ijalb3"
        fill="saltBox"
      />
      <circle
        className="css-1ijalb3"
        fill="mischka"
      />
      <g
        className="css-1itd6tp"
      >
        <text
          className="css-1i93ua"
          color="color"
          fill="outerSpace"
          textAnchor="middle"
          x={NaN}
          y={NaN}
        />
      </g>
    </svg>
  </div>
  <div
    className="css-0"
    position="relative"
  >
    <div
      className="css-1cncp8y"
      position="absolute"
    >
      <span
        backgroundColor="mischka"
        className="css-1bxc77d"
        color="outerSpace"
        position="relative"
      >
        <div
          backgroundColor="mischka"
          className="css-34aj3n"
          position="absolute"
        />
        <div
          className="css-4obz46"
          position="relative"
        >
          Take a break!
        </div>
      </span>
    </div>
  </div>
</div>
`;
